src/web100srv.c

```diff
@@ -1951,375 +1837,335 @@ void NDT_server_main_loop(SSL_CTX *ssl_context, int listenfd) {
-      // loop back to the begining.  There would need to be some
-      // additional logic installed if this AAA test relied on more than
-      // the client's IP address.  The client would also require
-      // modification to allow more credentials to be created/passed
+      // At this point we have received a connection from a client, meaning that
+      // a test is being requested.  At this point we should apply any policy or
+      // AAA functions to the incoming connection.  If we don't like the client,
+      // we can refuse the connection and loop back to the begining.  There
+      // would need to be some additional logic installed if this AAA test
+      // relied on more than the client's IP address.  The client would also
+      // require modification to allow more credentials to be created/passed
       // between the user and this application.
     }
 
-    if (pipe(chld_pipe) == -1)
+    if (pipe(chld_pipe) == -1) {
       log_println(6, "pipe() failed errno=%d", errno);
-    chld_pid = fork();
+      close(ctlsockfd);
+      continue;
+    }
 
-    switch (chld_pid) {
-      case -1: /* an error occured, log it and quit */
-        log_println(0, "fork() failed, errno = %d", errno);
-        /* todo: handle error and continue */
-        break;
-      default:  // this is the parent process, handle scheduling and
-                // queuing of multiple incoming clients
-
-        log_println(5, "Parent process spawned child = %d", chld_pid);
-        log_println(5, "Parent thinks pipe() returned fd0=%d, fd1=%d",
-                    chld_pipe[0], chld_pipe[1]);
-
-        close(chld_pipe[0]);
-
-        // Check to see if we have more than max_clients waiting in the
-        // queue
-        //  If so, tell them to go away.
-        //  changed for M-Lab deployment  1/28/09  RAC
-
-        if (((multiple == 0) && (waiting >= (max_clients - 1))) ||
-            ((multiple == 1) && (waiting >= ((4 * max_clients) - 1)))) {
-          log_println(
-              0,
-              "Too many clients/mclients (%d) waiting to be served, "
-              "Please try again later.", chld_pid);
-          send_json_message(ctlsockfd, SRV_QUEUE,
-            SRV_QUEUE_SERVER_BUSY_STR, 
-            strlen(SRV_QUEUE_SERVER_BUSY_STR),
-            testopt.connection_flags, JSON_SINGLE_VALUE);
-          close(chld_pipe[0]);
-          close(chld_pipe[1]);
-          shutdown(ctlsockfd, SHUT_WR);
-          close(ctlsockfd);
-          kill(chld_pid, SIGTERM);
-          if (new_child != NULL) {
-            log_println(6, "Too many clients freeing child=0x%x",
-                        new_child);
-            free(new_child);
-          }
-          continue;
+    chld_pid = fork();
+    if (chld_pid == -1) {
+      // an error occured, log it and keep waiting
+      log_println(0, "fork() failed, errno = %d (%s)", errno, strerror(errno));
+      close(child_pipe[0]);
```

Comment:
I don't understand `child_pipe`. How are we writing to it without ever initializing it?

```
$grep -R -n "child_pipe" src/* --include=*.[ch]
src/web100srv.c:1574:  int child_pipe[2];
src/web100srv.c:1876:      close(child_pipe[0]);
src/web100srv.c:1877:      close(child_pipe[1]);
src/web100srv.c:1917:        write(child_pipe[1], &child_signal, sizeof(int));
src/web100srv.c:1922:        close(child_pipe[1]);
src/web100srv.c:1959:        write(child_pipe[1], &child_signal, sizeof(int));
src/web100srv.c:2008:        write(child_pipe[1], &child_signal, sizeof(int));
src/web100srv.c:2018:        write(child_pipe[1], &child_signal, sizeof(int));
src/web100srv.c:2050:          write(child_pipe[1], &child_signal, sizeof(int));
```